src/ATen/function_wrapper.py

```diff
@@ -286,6 +305,13 @@ def insert(argument):
                 insert(argument)
             elif is_real_argument_to_wrapper(argument):
                 insert(argument)
+        if any(has_output_mask(arg) for arg in option['arguments']):
+            mask_size = sum(has_output_mask(arg) for arg in option['arguments'])
+            insert({
+                'name': 'output_mask',
+                'type': 'std::array<bool, {}>'.format(mask_size),
```

Comment:
We don't currently support `out=...` arguments in the automatic differentiation API. (And the backwards functions also have to be differentiable).

It tricky to support output arguments because the arguments have to be defined, but their history get's overwritten by the call. We also have to make sure they're not a view on another tensor and that their version counter is updated. We don't allow in-place operations on leaf-variables and these variables would necessarily be leaves.

Reply:
I'd also add that using return-values instead of optional output arguments makes for a much nicer API. The calls to convXXX_backward in PyTorch are a lot cleaner after these changes:

[convolution.diff](https://gist.github.com/colesbury/159a4347e0bb1e4e6a38a16bda3ef8ad)